nox-utils/src/tmlt/nox_utils/_release.py

```diff
@@ -0,0 +1,167 @@
+    repo.git.fetch("origin", "--tags")
+
+    if RemoteReference(repo, f"refs/remotes/origin/release/{version}") in origin.refs:
+        sess.error(f"Branch release/{version} already exists")
+    if Tag(repo, f"refs/tags/{version}") in Tag.list_items(repo):
+        sess.error(f"Release tag {version} already exists")
+
+    if not is_prerelease:
+        sess.debug('Updating changelog "Unreleased" header...')
+        try:
+            _update_changelog_unreleased(version)
+        except RuntimeError as e:
+            sess.error(str(e))
+    else:
+        sess.debug("This is a prerelease, skipping changelog update")
+
+    sess.debug("Creating release branch...")
+    release_branch = repo.create_head(f"release/{version}")
+    repo.head.reference = release_branch  # type: ignore[misc]
+    sess.debug("Creating release commit...")
+    repo.index.add(["CHANGELOG.rst"])
+    repo.index.commit(f"[auto] Release {version}")
+    sess.debug("Creating release tag...")
+    release_tag = repo.create_tag(version)
+
+    if not is_prerelease:
+        _add_changelog_unreleased()
+
+    sess.debug("Creating post-release commit...")
+    repo.index.add(["CHANGELOG.rst"])
+    repo.index.commit(f"[auto] Post-release {version}")
+
+    sess.debug("Pushing release tag and branch...")
+    origin.push(release_branch)
+    origin.push(release_tag)
+
+    sess.debug("Switching back to original branch...")
+    repo.head.reference = current_branch  # type: ignore[misc]
+    repo.head.reset(index=True, working_tree=True)
+
+    sess.debug(
+        "Release push completed, tag is at: "
+        f"https://github.com/{package_github}/releases/tag/{release_tag}"
+    )
+
+    sess.debug(
+        "For non-dev releases, please merge release branch to main by creating a PR "
+        "(even if there are no changes):"
+    )
+    sess.debug(f"https://github.com/{package_github}/compare/{release_branch}?expand=1")
+
+
+def _update_changelog_unreleased(version: str) -> None:
+    """Replace the "Unreleased" changelog header with one for the given version.
+
+    Also adds an anchor for the version number so we can reference it elsewhere.
+    The automatic anchor for the header doesn't work because Sphinx doesn't
+    support anchors that start with digits.
+    """
+    with Path("CHANGELOG.rst").open("r", encoding="utf-8") as fp:
+        changelog_content = fp.readlines()
+    for i, content in enumerate(changelog_content):
+        if re.match("^Unreleased$", content):
+            # BEFORE
+            # Unreleased
+            # ----------
+
+            # AFTER
+            # .. _v1.2.3:
+            #
+            # 1.2.3 - 2020-01-01
+            # ------------------
+            version_header = f"{version} - {datetime.date.today()}"
+            anchor = f".. _v{version}:\n\n"
+            subsection = f"{version_header}\n{'-' * len(version_header)}\n"
+
+            changelog_content[i] = anchor
+            changelog_content[i + 1] = subsection
```

Comment:
For your consideration (but not required): it's a little counter-intuitive to cram multiple lines into each entry in the content array. It might be a little more natural to remove the old lines and insert the new ones into the array?